File: NCServer/src/ncserver/connection/HttpSession.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package ncserver.connection;

import com.sun.net.httpserver.Headers;
import java.util.HashMap;
import java.util.concurrent.atomic.AtomicReference;
import ncserver.utils.DataReader;

/**
 *
 * @author Vladimir
 */
public abstract class HttpSession extends AbstractSession implements IHttpSession {

    // <editor-fold defaultstate="collapsed" desc="Реализация абстрактного AbstractSession">        
    @Override
    public boolean receiveData(Object data, int size) {
        // Принимаем данные - копируем данные в буфер            
        DataReader dataReader = new DataReader((byte[])data, 0, size);

        if (dataReader.getDataSize() < 3) {
            onReceivingError("Internal error: no message");
            return false;
        }
        String command = "";
        AtomicReference<String> commandRef = new AtomicReference<>(command);
        dataReader.readASCII(commandRef);
        command = commandRef.get();
        int headerSize = command.length();

        MessageHandler handler = inMessageHandlers.get(command);

        // Проверяем есть ли обработчик для сообщения такого типа
        if (handler == null) {
            onReceivingError("Internal error: unknown command");
            return false;
        }

        if (dataReader.getDataSize() < headerSize + 2) {
            onReceivingError("Internal error: no data");
            return false;
        }

        // сообщения с переменным размером - длину сообщения определяем из поля <размер данных>
        byte[] dataBuffer = new byte[dataReader.getDataSize() - dataReader.getPosition()];
        AtomicReference<byte[]> dataBufferRef = new AtomicReference<>(dataBuffer);
        dataReader.readArray(dataBufferRef);
        try {
            // Обрабатываем сообщение - вызываем соответствующий обработчик                        
            return handler.call(null, dataBufferRef.get(), dataBuffer.length);
        }
        catch (Exception exc) {
            String s = String.format("Internal error %s", exc.getMessage());
            onReceivingError(s);
            return false;
        }
    }
    // </editor-fold>

    // <editor-fold defaultstate="collapsed" desc="Реализация интерфейса IHttpSession">     
    @Override
    public boolean receiveData(String command, Headers headers, Object data, int size) {
        MessageHandler handler = inMessageHandlers.get(command);
        // Проверяем есть ли обработчик для сообщения такого типа
        if (handler == null) {
            onReceivingError("Internal error: unknown command");
            return false;
        }
        try {
            // Обрабатываем сообщение - вызываем соответствующий обработчик                        
            return handler.call(headers, (byte[])data, size);
        }
        catch (Exception exc) {
            String s = String.format("Internal error %s", exc.getMessage());
            onReceivingError(s);
            return false;
        }
    }
    // </editor-fold>

    public HttpSession(IConnectManager connManager, IConnectionInfo connection) {
        connectManager = connManager;
        connectInfo = connection;
    }

    protected abstract class MessageHandler {
        protected abstract boolean call(Headers headers, byte[] data, int size);
    }

    private final HashMap<Object, MessageHandler> inMessageHandlers = new HashMap<Object, MessageHandler>();

    protected final void addMessageHandler(Object type, MessageHandler messageHandlers) {
        inMessageHandlers.put(type, messageHandlers);
    }

    protected void onReceivingError(String errorMsg) {
        System.err.println(errorMsg);
    }

    @Override
    protected void onSessionClosing() {
        super.onSessionClosing();
        inMessageHandlers.clear();
    }
}

File: NoughtsAndCrosses/Connection/HTTP/HttpSession.cs
using System;
using System.Collections.Generic;
using NoughtsAndCrosses.Utils;

namespace NoughtsAndCrosses.Connection.HTTP {
  public class HttpSession : Session {

    /// <summary>
    /// Делегат обработчика сообщения
    /// </summary>
    /// <param name="data">Массив с данными</param>
    /// <param name="offset">Смещение начала данных</param>
    /// <param name="size">Размер данных</param>
    protected delegate bool MessageHandler(byte[] data, int offset, int size);

    private readonly Dictionary<object, MessageHandler> inMessageHandlers = new Dictionary<object, MessageHandler>();

    public HttpSession(IConnectManager connectHandler, IConnectionInfo connection)
    : base(connectHandler, connection) {

    }

    protected void addMessageHandler(object type, MessageHandler messageHandlers) {
      inMessageHandlers.Add(type, messageHandlers);
    }

    #region Прием сообщений

    /// <summary>
    /// Прием данных
    /// </summary>
    /// <param name="data">данные</param>
    /// <param name="size">размер данных</param>
    public sealed override void ReceiveData(byte[] data, int size) {
      // Принимаем данные - копируем данные в буфер
      DataReader dataReader = new DataReader(data, 0, size);

      if (dataReader.GetDataSize() < 3) {
        OnReceivingError(RECEIVE_FATAL_ERROR, "Internal error: no message");
        return;
      }
      // Формируем сообщение
      // по сигнатуре проверяем достоверность полученного сообщения

      // Определяем тип сообщения
      byte type = 255;
      dataReader.Read(ref type);

      if (!inMessageHandlers.ContainsKey(type)) {
        OnReceivingError(RECEIVE_FATAL_ERROR, "Internal error: unknown command");
        return;
      }

      int headerSize = 1;

      // Проверяем есть ли обработчик для сообщения такого типа
      if (dataReader.GetDataSize() < headerSize + 2) {
        OnReceivingError(RECEIVE_FATAL_ERROR, "Internal error: no data");
        return;
      }

      // сообщения с переменным размером - длину сообщения определяем из поля <размер данных>
      byte[] dataBuffer = new byte[dataReader.GetDataSize() - dataReader.GetPosition()];
      dataReader.ReadArray(ref dataBuffer);
      try {
        // Обрабатываем сообщение - вызываем соответствующий обработчик
        inMessageHandlers[type](dataBuffer, headerSize, dataBuffer.Length);
      }
      catch (Exception exc) {
        string s = string.Format("Internal error {0}", exc.Message);
        OnReceivingError(Session.RECEIVE_FATAL_ERROR, s);
        return;
      }
    }

    public sealed override void ReceiveData(DataBuffer buffer) {
      ReceiveData(buffer.GetBuffer(), (int)buffer.GetBufferSize());
    }

    public override void ReceiveData(string command, byte[] data, int size) {
      if (!inMessageHandlers.ContainsKey(command)) {
        OnReceivingError(RECEIVE_FATAL_ERROR, "Internal error: unknown command");
        return;
      }
      try {
        // Обрабатываем сообщение - вызываем соответствующий обработчик
        inMessageHandlers[command](data, 0, size);
      }
      catch (Exception exc) {
        string s = string.Format("Internal error {0}", exc.Message);
        OnReceivingError(Session.RECEIVE_FATAL_ERROR, s);
        return;
      }
    }

    #endregion

    #region Отправка сообщения

    public void SendData(string method, string command, Headers headers, DataBuffer data) {
      if (connectHandler == null) {
        return;
      }
      if (connectInfo == null) {
        return;
      }
      if (IsClosed()) {
        return;
      }
      (connectHandler as IHttpClient).SendData(connectInfo, method, command, headers, data);
    }

    public void SendJsonData(string method, string command, Headers headers, string json) {
      if (connectHandler == null) {
        return;
      }
      if (connectInfo == null) {
        return;
      }
      if (IsClosed()) {
        return;
      }
      (connectHandler as IHttpClient).SendJson(connectInfo, method, command, headers, json);
    }

    #endregion
  }
}
